### PR TITLE
Bug 1834136: fix(queues): use a single gc queue

### PR DESCRIPTION
**Pull request by njhale**
**Description of the change:**

- Use a single GC queue for notifying CSVs of changes to resources that affect them
- Requeue events all-namespace queues if registered

### CODE REVIEW

File: pkg/controller/operators/olm/operator.go [hunk 1]
```diff
@@ -702,13 +702,14 @@ func (a *Operator) syncObject(obj interface{}) (syncError error) {
...
 			switch metaObj.(type) {
-			case *rbacv1.ClusterRole, *rbacv1.ClusterRoleBinding:
+			case *rbacv1.ClusterRole, *rbacv1.ClusterRoleBinding, *admissionregistrationv1.MutatingWebhookConfiguration, *admissionregistrationv1.ValidatingWebhookConfiguration:
 				resourceEvent := kubestate.NewResourceEvent(
 					kubestate.ResourceUpdated,
 					metaObj,
 				)
-				syncError = a.objGCQueueSet.RequeueEvent(ns, resourceEvent)
-				logger.Debugf("syncObject - requeued update event for %v, res=%v", resourceEvent, syncError)
+				if syncError = a.objGCQueueSet.RequeueEvent("", resourceEvent); syncError != nil {
```
Comment (njhale) [inline pkg/controller/operators/olm/operator.go:710]: Sanity check: Does this requeue indefinitely?

File: pkg/controller/operators/olm/operator.go [hunk 2]
```diff
@@ -950,36 +951,42 @@ func (a *Operator) handleClusterServiceVersionDeletion(obj interface{}) {
...
+		}
 	}
 
 	webhookSelector := labels.SelectorFromSet(ownerutil.OwnerLabel(clusterServiceVersion, v1alpha1.ClusterServiceVersionKind)).String()
 	mWebhooks, err := a.opClient.KubernetesInterface().AdmissionregistrationV1().MutatingWebhookConfigurations().List(context.TODO(), metav1.ListOptions{LabelSelector: webhookSelector})
 	if err != nil {
 		logger.WithError(err).Warn("cannot list MutatingWebhookConfigurations")
 	}
-	for _, webhook := range mWebhooks.Items {
-		syncError := a.objGCQueueSet.RequeueEvent("", kubestate.NewResourceEvent(kubestate.ResourceUpdated, &webhook))
```
Comment (njhale) [inline pkg/controller/operators/olm/operator.go:972]: If I'm not mistaken, the value of `webhook` was changing underneath the pointer since it's a free variable in the loop's body.

**Comment by awgreene** [comment]
/hold
Waiting for @dinhxuanvu or @ecordell to get a second set of eyes on this.

File: pkg/controller/operators/olm/operator.go [hunk 3]
```diff
@@ -950,36 +951,42 @@ func (a *Operator) handleClusterServiceVersionDeletion(obj interface{}) {
...
-		logger.Debugf("handleCSVdeletion - requeued update event for %v, res=%v", webhook, syncError)
+	for i := 0; i < len(mWebhooks.Items); i++ {
```
Comment (ecordell) [inline pkg/controller/operators/olm/operator.go:974]: you can still keep the `range`, just make sure to do `w := webhook` in the loop body

File: pkg/lib/queueinformer/resourcequeue.go [hunk 4]
```diff
@@ -39,6 +39,11 @@ func (r *ResourceQueueSet) RequeueEvent(namespace string, resourceEvent kubestat
 	r.mutex.RLock()
 	defer r.mutex.RUnlock()
 
+	if queue, ok := r.queueSet[metav1.NamespaceAll]; len(r.queueSet) == 1 && ok {
```
Comment (ecordell) [inline pkg/lib/queueinformer/resourcequeue.go:42]: Why is this necessary? if `""` is passed as the namespace, the following block will requeue in the "allnamespace" queue

File: pkg/controller/operators/olm/operator.go (hunk 2 above)
Comment (awgreene) [reply pkg/controller/operators/olm/operator.go:972]: Great catch!

File: pkg/lib/queueinformer/resourcequeue.go (hunk 4 above)
Comment (njhale) [reply pkg/lib/queueinformer/resourcequeue.go:42]: because we also want all of the requeues that specifically mention a namespace to go to the all-namespace queue as well.
